File: FrontendMockApp/Code Examples/Heavy/TimeoutExample.cs
//recursive solution of the knapsack problem (should timeout because of O(n^2))
class TimeoutExample
{
    public class Solution
    {
        public double FractionalKnapsack(Item[] items, int capacity)
        {            
            var sortedItems = items.OrderByDescending(i => i.Ratio).ToArray();

            return FractionalKnapsackRecursive(sortedItems, capacity, 0);
        }

        private double FractionalKnapsackRecursive(Item[] items, int capacity, int index)
        {
            if (capacity == 0 || index == items.Length)
                return 0.0;

            var current = items[index];

            if (current.Weight <= capacity)
            {
                return current.Value + FractionalKnapsackRecursive(items, capacity - current.Weight, index + 1);
            }
            else
            {
                return current.Ratio * capacity;
            }
        }
    }
}

File: FrontendMockApp/Code Examples/Heavy/TimeoutExample.kt
class Solution {
    fun fractionalKnapsack(items: Array<Item>, capacity: Int): Double {
        val sortedItems = items.sortedByDescending { it.ratio }.toTypedArray()
        return fractionalKnapsackRecursive(sortedItems, capacity, 0)
    }

    // Artificially exponential version (не оптимальный, просто ради нагрузки)
    private fun fractionalKnapsackRecursive(items: Array<Item>, capacity: Int, index: Int): Double {
        if (capacity == 0 || index == items.size) return 0.0

        var takeFull = 0.0
        if (items[index].weight <= capacity) {
            takeFull = items[index].value + fractionalKnapsackRecursive(items, capacity - items[index].weight, index + 1)
        }

        val takePartial = items[index].ratio * minOf(capacity, items[index].weight)

        val skip = fractionalKnapsackRecursive(items, capacity, index + 1)

        return maxOf(maxOf(takeFull, takePartial), skip)
    }
}
